Fix race accessing the connection proxy state in finalize()

julioc-lopez opened the pull request:
Accesses to ConnectionProxy.state are unprotected and thus racy. This
leads to a situation where reading 'state' in the finalizer thread
gets an stale value long after it has been set to CLOSED in
ConnectionProxy.invoke in a different thread. This behavior has
been observed in unit tests for products that rely on JDBC lint.

Uses an atomic reference to ensure memory visibility. The
happens-after relationship is ensured by the fact that the
finalizer only runs after there are no reachable references
to the object. This means, this happens after the invoke
method has been executed.

Commit:
Fix race accessing the connection proxy state in finalize()

Accesses to ConnectionProxy.state are unprotected and thus racy. This
leads to a situation where reading 'state' in the finalizer thread
gets an stale value long after it has been set to CLOSED in
ConnectionProxy.invoke in a different thread. This behavior has
been observed in unit tests for products that rely on JDBC lint.

Uses an atomic reference to ensure memory-visibility. The
happens-after relationship is ensured by the fact that the
finalizer only runs after there are no reachable references
to the object. This means, this happens after the invoke
method has been executed.

julio-maginatics commented on src/main/java/com/maginatics/jdbclint/ConnectionProxy.java:
```diff
@@ -124,7 +127,7 @@ public Object invoke(final Object proxy, final Method method,
     @Override
     protected void finalize() throws SQLException {
         if (config.isEnabled(Check.CONNECTION_MISSING_CLOSE) &&
-                state != State.CLOSED) {
```
`state` can get a stale value here

andrewgaul commented:
Do `StatementProxy`, `ResultSetProxy`, and `BlobProxy` need corresponding changes?

julio-maginatics commented:
Good question.

Let me check quickly, I didn't look into it.

julio-maginatics commented:
Yeap. I'll submit separate patches for each of those.

Commit:
Ensure memory-visibility for BlobProxy.freed in finalize()

See PR #23

Commit:
Ensure memory visibility for StatementProxy.state in finalize()

See PR #23

Commit:
Ensure memory visibility for ResultSetProxy.closed in finalize()

See PR #23

Commit:
Prefer compareAndSet to set BlobProxy.freed

Even though it is not strictly required since a blob is expected to be
used only a single thread and not passed across threads.

See PR #23

andrewgaul commented on src/main/java/com/maginatics/jdbclint/BlobProxy.java:
```diff
@@ -55,10 +56,10 @@ public Object invoke(final Object proxy, final Method method,
         String name = method.getName();
         if (name.equals("free")) {
-            if (config.isEnabled(Check.BLOB_DOUBLE_FREE) && freed) {
+            boolean previouslyFreed = !freed.compareAndSet(false, true);
```
Simpler to call `getAndSet`?

julio-maginatics replied:
yeap

julio-maginatics replied:
will add a patch at the end of the train to avoid rebasing

andrewgaul commented on src/main/java/com/maginatics/jdbclint/ResultSetProxy.java:
```diff
@@ -94,10 +95,11 @@ public Object invoke(final Object proxy, final Method method,
         String name = method.getName();
         if (name.equals("close")) {
-            if (config.isEnabled(Check.RESULT_SET_DOUBLE_CLOSE) && closed) {
+            boolean previouslyClosed = !closed.compareAndSet(false, true);
```
Same as above.

julio-maginatics replied:
done

Commit:
Prefer compareAndSet to set ConnectionProxy.state

Commit:
Cleanup: fix indentation

Commit:
Return after closing {Connection,Statement,ResultSet} in proxies

This avoids invoking close() twice in the underlying objects.

Commit:
Prefer AtomicBoolean.getAndSet()

Uses getAndSet() for BlobProxy.freed and ResultSetProxy.closed.
It is simpler, and more importantly easier to understand, in these
use cases.

PR #23

julio-maginatics commented:
Andrew: Please let me know if you have additional comments. I'm planning to merge this and generate a new release.

gaul reviewed (approved):
Please squash.
